kong/clustering/wrpc_data_plane.lua

```diff
@@ -0,0 +1,374 @@
+  end
+
+  if declarative.get_current_hash() == new_hash then
+    ngx_log(ngx_DEBUG, _log_prefix, "same config received from control plane, ",
+                                    "no need to reload")
+    return true
+  end
+
+  -- NOTE: no worker mutex needed as this code can only be
+  -- executed by worker 0
+  local res
+  res, err = declarative.load_into_cache_with_events(entities, meta, new_hash)
+  if not res then
+    return nil, err
+  end
+
+  if update_cache then
+    -- local persistence only after load finishes without error
+    local f
+    f, err = io_open(CONFIG_CACHE, "w")
+    if not f then
+      ngx_log(ngx_ERR, _log_prefix, "unable to open config cache file: ", err)
+
+    else
+      local config = assert(cjson_encode(config_table))
+      config = assert(self:encode_config(config))
+      res, err = f:write(config)
+      if not res then
+        ngx_log(ngx_ERR, _log_prefix, "unable to write config cache file: ", err)
+      end
+
+      f:close()
+    end
+  end
+
+  return true
+end
+
+
+function _M:init_worker()
+  -- ROLE = "data_plane"
+
+  if ngx.worker.id() == 0 then
+    local f = io_open(CONFIG_CACHE, "r")
+    if f then
+      local config, err = f:read("*a")
+      if not config then
+        ngx_log(ngx_ERR, _log_prefix, "unable to read cached config file: ", err)
+      end
+
+      f:close()
+
+      if config and #config > 0 then
+        ngx_log(ngx_INFO, _log_prefix, "found cached config, loading...")
+        config, err = self:decode_config(config)
+        if config then
+          config, err = cjson_decode(config)
+          if config then
+            local res
+            res, err = self:update_config(config)
+            if not res then
+              ngx_log(ngx_ERR, _log_prefix, "unable to update running config from cache: ", err)
+            end
+
+          else
+            ngx_log(ngx_ERR, _log_prefix, "unable to json decode cached config: ", err, ", ignoring")
+          end
+
+        else
+          ngx_log(ngx_ERR, _log_prefix, "unable to decode cached config: ", err, ", ignoring")
+        end
+      end
+
+    else
+      -- CONFIG_CACHE does not exist, pre create one with 0600 permission
+      local flags = bit.bor(system_constants.O_RDONLY(),
+                            system_constants.O_CREAT())
+
+      local mode = ffi.new("int", bit.bor(system_constants.S_IRUSR(),
+                                          system_constants.S_IWUSR()))
+
+      local fd = ffi.C.open(CONFIG_CACHE, flags, mode)
+      if fd == -1 then
+        ngx_log(ngx_ERR, _log_prefix, "unable to pre-create cached config file: ",
+                ffi.string(ffi.C.strerror(ffi.errno())))
+
+      else
+        ffi.C.close(fd)
+      end
+    end
+
+    assert(ngx.timer.at(0, function(premature)
+      self:communicate(premature)
+    end))
+  end
+end
+
+
+local wrpc_config_service
+local function get_config_service()
+  if not wrpc_config_service then
+    wrpc_config_service = wrpc.new_service()
```

Comment:
do all ngnix workers need to share the same service or each worker can create its own?

Reply:
I need to do something like
```
    self.peer = wrpc.new_remote_client(get_requests_log_service(), self.TELEMETRY_CHANNEL_NAME)
```

Reply:
My understanding is that `local` means each worker will create its own `wrpc_requests_log_service`

Reply:
> ... share the same service ...
the service is a Lua object.  as such, it can't be shared between workers.

yes, each worker will need its own, like everything in openresty.